Remove duplicate RunCommand functions used for integration-cli

dnephin opened the pull request:
There are currently ~20 different functions that do essentially the same thing (running a command). Many of these are just single line wrappers around other commands. 

This PR introduces a new function (`RunCmd()`) which:
- handles all variations of the previous ~20 functions
- provides a way to assert the expected result of the command, with a complete error message on failure

Also:
- removes many (12) of the lesser used function and replaces them with direct calls to `RunCmd()`
- updates the remaining functions to be thin wrappers around the new `RunCmd()`, and use `Result.Assert()` to provide better error messages
- removes `unmarshalJSON()` which was just an alias for `json.Unmarshal`

Once this PR is merged I can continue to convert some of the lesser used functions to direct calls to `RunCmd()` to further reduce the number of functions.

The "new" canonical way looks something like this:

``` go

// Expect the command to be successful 
result := icmd.RunCommand("echo hello world")
result.Assert(c, icmd.Expected{})

// Expect the command to fail
result := icmd.RunCommand("bash -c 'echo oops; exit 3'")
result.Assert(c, icmd.Expected{
    ExitCode: 3,
    Out: "oops",
})
```

If you need to set anything on the command (directory, env, timeout, stdin, stdout), you can use the `RunCmd()` call and pass a `Cmd{}` struct which accepts all of these options.

dnephin committed:
Remove unnecessary json.Unmarshal wrapper.

Signed-off-by: Daniel Nephin <dnephin@docker.com>

dnephin committed:
Create a unified RunCommand function with Assert()

Remove some run functions and replace them with the unified run command.
Remove DockerCmdWithStdoutStderr
Remove many duplicate runCommand functions.
Also add dockerCmdWithResult()
Allow Result.Assert() to ignore the error message if an exit status is expected.
Fix race in DockerSuite.TestDockerInspectMultipleNetwork
Fix flaky test DockerSuite.TestRunInteractiveWithRestartPolicy

Signed-off-by: Daniel Nephin <dnephin@docker.com>

aaronlehmann commented:
I'm not really a fan of using shlex for this. The old helper functions took a variable number of arguments, one for each command argument. I thought this worked well. Having the helper lex the command line and split the arguments will be problematic if/when some args contain spaces or other metacharacters. Code authors will have to be careful to escape things in many places as if they were writing shell scripts. Also, I expect we'll end up with a lot of places where we use string concatentation or `fmt.Sprintf` to assemble commands from individual arguments, only to have them split into args again by shlex. For example, I see that `exec.Command(dockerBinary, "unpause", container)` became `return icmd.RunCommand(dockerBinary + " unpause " + container).Error`. I find the former a lot cleaner.

It's good that there's a mode that allows the arguments to be explicitly listed, but what's the value of supporting shell lexing at all? I'm afraid it will encourage mistakes and bad code.

dnephin commented:
>  The old helper functions took a variable number of arguments, one for each command argument. I thought this worked well.

You can still do that with `CommandArgs`. It works, but it can make the tests a lot harder to read. Using shlex moves us closer to a more bdd style of test, where the test is actually readable.

> Having the helper lex the command line and split the arguments will be problematic if/when some args contain spaces or other metacharacters

I don't understand this. That is what quotes are for. It lets the tests read the same way as you would run the command on the command line.

>  I expect we'll end up with a lot of places where we use string concatenation or `fmt.Sprintf` to assemble commands

Ya, right now it's kind of a mix. There are cases when it's a lot more readable to have just a single string, and there are cases where it doesn't make any improvement (like in the example you found).  I tried to support both.

dnephin commented:
This is now passing all suites (except for win2lin which I think is still broken?).

I've squashed a bunch of the commits

cpuguy83 commented:
LGTM

tonistiigi commented:
I'm also not really a fan of shlex here. There are only a handful of places where it has been added so the usage is not consistent. Half of the places where it's added use string concatenation with added spaces or `sprintf`. And for some reason just seems so unsafe to me. Are there more maintainers who like shlex here and see it as a good direction for our tests?

dnephin commented:
If no one else likes the shlex I can remove it. It could also be added as a helper method for the few cases where it is an improvement, and keep the default the args list.

dnephin committed:
Change to use c.Assert() instead of result.Assert()

Fix delete containers and make sure it prints errors correctly.
Rename Result.Fails to Result.Assert()
Create a constant for the default expected.

Signed-off-by: Daniel Nephin <dnephin@docker.com>

dnephin commented:
Ok, shlex is gone, and the build is finally green!

This is ready for another look.

dnephin commented on integration-cli/docker_cli_attach_unix_test.go:
```diff
@@ -80,7 +79,11 @@ func (s *DockerSuite) TestAttachAfterDetach(c *check.C) {
+			buff := make([]byte, 200)
+			tty.Read(buff)
+			c.Fatalf("%s: %s", err, buff)
+		}
```
This was just to improve the error message as I was debugging the test failure. I can remove it, but I thought I'd leave it in case someone else hits a failure here.

vdemeester commented on integration-cli/docker_utils.go:
```diff
@@ -464,49 +441,70 @@ func dockerCmdWithError(args ...string) (string, int, error) {
+	return icmd.RunCommand(dockerBinary, args...)
+}
+
+func binaryWithArgs(args []string) []string {
```
why this one doesn't take `...string` ? 👼

dnephin replied:
hmm, I'm not sure. I guess it would be easier to use as a `...string`. I'll make that change

vdemeester commented on integration-cli/docker_utils.go:
```diff
@@ -228,16 +228,9 @@ func readBody(b io.ReadCloser) ([]byte, error) {
-	return err
+func deleteContainer(container ...string) error {
+	result := icmd.RunCommand(dockerBinary, append([]string{"rm", "-fv"}, container...)...)
+	return result.Fails(icmd.Expected{})
```
This feels a littly tricky to read 😅

dnephin replied:
Hmm, what is tricky about it? The first line, the second or both?

vdemeester replied:
The 2nd one :angel: `icmd.Expected{}` by itself in `result.Fails` doesn't read easily.

dnephin replied:
Hmm, ya `Fails()` is probably the wrong name.

Maybe `Compare()` or `Check()` ?

vdemeester replied:
Hum yep `Compare` might be more appropriated :angel:

tonistiigi commented:
Maybe instead of `result.Assert(c, icmd.Expected{})` `c.Assert(result, icmd.Check, optReq)` is more readable? Although most of the cases you only check error so maybe there aren't even lots benefits for having a special function.

vdemeester commented on pkg/integration/cmd/command.go:
```diff
@@ -0,0 +1,267 @@
+	// If a non-zero exit code is expected there is going to be an error.
+	// Don't require an error message as well as an exit code because the
+	// error message is going to be "exit status <code> which is not useful
+	case exp.Error == "" && exp.ExitCode != 0:
```
Just for the sake of reading this, I would have put the comment inside the case 👼

dnephin replied:
Is that normal? I always expect comments above the line they are commenting. I actually tried putting it inside the case while I was working on this, but it looked wrong to me, like it was commenting the line it was above, instead of the case.

vdemeester replied:
> I always expect comments above the line they are commenting

Same, except for switch case, even more when they are empty (the fact that `Go` doesn't fall through but lot's of languages does might confuse readers :sweat_smile: ).

dnephin commented on pkg/integration/cmd/command.go:
```diff
@@ -0,0 +1,267 @@
+}
+
+// Cmd is a command to run. One of Command or CommandArgs can be used to set the
+// comand line. Command will be paased to shlex and split into a string slice.
```
fixed

dnephin commented:
> Maybe instead of result.Assert(c, icmd.Expected{}) c.Assert(result, icmd.Check, optReq) is more readable? 

That would force it to be tied to `gocheck`. The way it is now it works with `testing.T`, or even outside of a test framework by using `result.Fails()`.

> Although most of the cases you only check error so maybe there aren't even lots benefits for having a special function.

I think you are missing one of the most important parts of `result.Assert(...)` which is the error message you get back. It contains the command, stdout/stderr/error, and the failed expectations. Without this it is much harder to debug these exec failures.

I think it would be possible to create a custom gocheck comparator, but it would be a lot more verbose, and it would only work within a go-check suite, which is kind of an unnecessary restriction.

dnephin commented:
I guess I could make the `Expected` use the gocheck comparator interface, so it could be written as

```go
c.Assert(result, icmd.Expected{...})
```

If that reads better, but it seems pretty similar to me.

tonistiigi commented:
> That would force it to be tied to gocheck. The way it is now it works with testing.T, or even outside of a test framework by using result.Fails().

The decision base integration tests on `gocheck`  was made a long time ago(I had no part in that btw). We should embrace it instead of having different styles in our tests.

> I think you are missing one of the most important parts of result.Assert(...) which is the error message you get back. It contains the command, stdout/stderr/error, and the failed expectations. Without this it is much harder to debug these exec failures.

Checker function should let you define all the same outputs. I believe simple error checks usually also include the output in the last parameter atm.

> I guess I could make the Expected use the gocheck comparator

I find it better. Maybe `Expected` isn't a best name though.

dnephin commented:
Ok, I'll leave the existing interface for places where the `check.C` isn't available, and try adding a custom checker that outputs the same result. The `Expected` object is still going to be necessary for many cases, but I guess the default of "no error" can be done without it.

tonistiigi commented:
@dnephin Only place I can see `Assert/Fails` used outside gocheck is unit testing the same package that adds them. We probably don't need it for that, or it is even bad because if they fail you don't know if the condition or the assertion logic failed. We could probalby just check the public methods instead including the one that will be used by gocheck.

dnephin commented:
There are still cases where a method like `Fails()` (with a better name) are valuable, even if it's within the integration test suite. You might want to make some assertion about something outside of the exec command, or make a more complex assertion about output,  other than just "contains". 

When that assertion fails you want to be able to provide more context about the failure, which comes from `Fails()`.  But I think we do need to rename it, because it doesn't read well right now.

The `Assert()` I guess we can remove and replace with a checker.

dnephin commented:
I just looked into the [Checker interface](https://godoc.org/gopkg.in/check.v1#Checker) again, and I don't see any way to write out a custom error message. You don't even get access to the `check.C` object which would let you log some messages directly. Am I missing some way that I can print a custom messages from a checker?

I'm back to thinking that `result.Assert()` is the right way to do this. We can definitely improve the names for these functions to make it read better.

Some thoughts:
* rename `Fails()` to something like `Compare`, so it would read: `result.Compare(Expected)`
* create a constant for the default `Expected` object, so it would read `result.Assert(c, icmd.ExpectNoError)` 
* `Expected` could also be renamed to something else. I don't have any better ideas right now.

dnephin commented:
Ok, I've made the following changes:
* removed `result.Assert()` and added `Matches` which is a gocheck Checker
* created a constant for the default `Expected`: `c.Assert(result, icmd.Matches, icmd.Success)`
* Renamed `Fails()` to `Compare()`

tonistiigi commented:
LGTM.

Could you squash? Maybe to 2-3 commits.

dnephin commented:
Cool, squashed to 3 commits

vdemeester commented on integration-cli/docker_utils.go:
```diff
@@ -464,49 +441,70 @@ func dockerCmdWithError(args ...string) (string, int, error) {
-	return integration.DockerCmdWithStdoutStderr(dockerBinary, c, args...)
+
+	result := icmd.RunCommand(dockerBinary, args...)
+	// TODO: why is c ever nil?
```
I do agree with this comment :angel:, pretty sure the condition is not needed :angel:

dnephin replied:
Ya, I just copied the condition from the previous version, so I left the TODO.  I think it can be cleaned up in a future branch.

vdemeester replied:
yes :wink:

vdemeester commented:
LGTM 🐸
